usb/esp_tinyusb/tusb_cdc_acm.c

```diff
@@ -382,26 +382,34 @@ esp_err_t tusb_cdc_acm_init(const tinyusb_config_cdcacm_t *cfg)
     acm->ringbuf_read_mux = xSemaphoreCreateMutex();
     if (acm->ringbuf_read_mux == NULL) {
         ESP_LOGE(TAG, "Creation of a ringbuf mutex failed");
-        free_obj(itf);
-        return ESP_ERR_NO_MEM;
+        ret = ESP_ERR_NO_MEM;
+        goto clean;
     }
 
     acm->rx_tfbuf = malloc(CONFIG_TINYUSB_CDC_RX_BUFSIZE);
     if (!acm->rx_tfbuf) {
         ESP_LOGE(TAG, "Creation buffer error");
-        free_obj(itf);
-        return ESP_ERR_NO_MEM;
+        ret = ESP_ERR_NO_MEM;
+        goto clean;
     }
     acm->rx_unread_buf_sz = cfg->rx_unread_buf_sz == 0 ? RX_UNREADBUF_SZ_DEFAULT : cfg->rx_unread_buf_sz;
     acm->rx_unread_buf = xRingbufferCreate(acm->rx_unread_buf_sz, RINGBUF_TYPE_BYTEBUF);
     if (acm->rx_unread_buf == NULL) {
         ESP_LOGE(TAG, "Creation buffer error");
-        free_obj(itf);
-        return ESP_ERR_NO_MEM;
+        ret = ESP_ERR_NO_MEM;
+        goto clean;
     } else {
         ESP_LOGD(TAG, "Comm Initialized buff:%d bytes", cfg->rx_unread_buf_sz);
         return ESP_OK;
     }
+clean:
+    if (acm->rx_tfbuf) {
+        free(acm->rx_tfbuf);
+        acm->rx_tfbuf = NULL;
+    }
+    free_obj(itf);
+    tinyusb_cdc_deinit(itf);
```

Comment:
Could you please review the `free_obj` and `deinit` functions? I find it quite confusing. Could we merge the allocation and initialization (and analogously freeing and deinitialization) to just one function?

Reply:
esp_tusb_cdcacm_t is a subclass specific object (i.e. ACM) which is available only in tusb_sdc_acm.c. It is hidden from generic cdc class source file cdc.c.
alloc_obj()/free_obj() is for allocation/freeing the subclass specific object esp_tusb_cdcacm_t (Here, only ACM type) and thus used only in tusb_sdc_acm.c. By this design and segregation, we should not move the subclass specific data to generic cdc class file.

Reply:
@tore-espressif Please check and approve if no more comments.

Reply:
OK, agreed